### PR TITLE
fix(sbom,scan)!: BREAKING CHANGES, see description

### PR DESCRIPTION
**Pull request by luhring**
- ⚠️ **BREAKING**: The the API of several library methods associated with SBOM generation and vulnerability scanning have been modified in a breaking way.
- ⚠️ **BREAKING**: Scan result filtering with advisories now takes at most one advisories dir/index; multiple were never needed, so this simplifies the operation.
- 📜 Logging with clog and contexts has been plumbed extensively into SBOM generation and vulnerability scanning.
- 🔨  Fix: When the scan command fetched a remote APK, this temp file is now cleaned up after scanning, whereas it was not before.

cc: @cpanato @hectorj2f @rawlingsj @jonjohnsonjr

### CODE REVIEW

**Comment by luhring** [comment]
Example of new logging at DEBUG level:

```console
$ wolfictl scan -r atlantis -a ../wolfi-advisories -f all -vv > /dev/null
time=2024-04-25T10:17:47.651-04:00 level=INFO msg="scan results will be filtered using advisory data" filterSet=all advisoriesRepoDir=../wolfi-advisories
time=2024-04-25T10:17:48.427-04:00 level=DEBUG msg="downloading APK" url=https://packages.wolfi.dev/os/x86_64/atlantis-0.27.2-r3.apk
time=2024-04-25T10:17:48.867-04:00 level=INFO msg="downloaded APK" path=/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/x86_64-atlantis-0.27.2-r3-2452533565.apk
time=2024-04-25T10:17:49.077-04:00 level=DEBUG msg="downloading APK" url=https://packages.wolfi.dev/os/aarch64/atlantis-0.27.2-r3.apk
time=2024-04-25T10:17:49.670-04:00 level=INFO msg="downloaded APK" path=/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/aarch64-atlantis-0.27.2-r3-4098194326.apk
time=2024-04-25T10:17:49.688-04:00 level=DEBUG msg="checking cache for SBOM" expectedPath=/Users/dan/Library/Caches/wolfictl/sbom/apk/aarch64-atlantis-0.27.2-r3-4098194326-sha256-b815a50628ba7ebce22a9ac450e5b8cccacd3b5baa7778e9f5dbb1696b52c42f.syft.json
time=2024-04-25T10:17:49.688-04:00 level=DEBUG msg="SBOM cache miss" cachedPath=/Users/dan/Library/Caches/wolfictl/sbom/apk/aarch64-atlantis-0.27.2-r3-4098194326-sha256-b815a50628ba7ebce22a9ac450e5b8cccacd3b5baa7778e9f5dbb1696b52c42f.syft.json
time=2024-04-25T10:17:49.688-04:00 level=INFO msg="generating SBOM for APK file" path=/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/aarch64-atlantis-0.27.2-r3-4098194326.apk distroID=wolfi
time=2024-04-25T10:17:49.688-04:00 level=DEBUG msg="created temp directory to unpack APK" path=/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/wolfictl-sbom-1293666122
time=2024-04-25T10:17:49.689-04:00 level=DEBUG msg="checking cache for SBOM" expectedPath=/Users/dan/Library/Caches/wolfictl/sbom/apk/x86_64-atlantis-0.27.2-r3-2452533565-sha256-ff3b9bc755d38210bc577d34e4d84ad361f90a2ac24a35f869849b125af7ac69.syft.json
time=2024-04-25T10:17:49.689-04:00 level=DEBUG msg="SBOM cache miss" cachedPath=/Users/dan/Library/Caches/wolfictl/sbom/apk/x86_64-atlantis-0.27.2-r3-2452533565-sha256-ff3b9bc755d38210bc577d34e4d84ad361f90a2ac24a35f869849b125af7ac69.syft.json
time=2024-04-25T10:17:49.689-04:00 level=INFO msg="generating SBOM for APK file" path=/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/x86_64-atlantis-0.27.2-r3-2452533565.apk distroID=wolfi
time=2024-04-25T10:17:49.690-04:00 level=DEBUG msg="created temp directory to unpack APK" path=/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/wolfictl-sbom-3770624166
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="unpacked APK file to temp directory" apkFilePath=/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/aarch64-atlantis-0.27.2-r3-4098194326.apk
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="apk temp directory item" path=.
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="apk temp directory item" path=.PKGINFO
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="apk temp directory item" path=.SIGN.RSA.wolfi-signing.rsa.pub
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="apk temp directory item" path=usr
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="apk temp directory item" path=usr/bin
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="apk temp directory item" path=usr/bin/atlantis
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="apk temp directory item" path=var
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="apk temp directory item" path=var/lib
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="apk temp directory item" path=var/lib/db
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="apk temp directory item" path=var/lib/db/sbom
time=2024-04-25T10:17:49.942-04:00 level=DEBUG msg="apk temp directory item" path=var/lib/db/sbom/atlantis-0.27.2-r3.spdx.json
time=2024-04-25T10:17:49.943-04:00 level=DEBUG msg="synthesized APK package for SBOM" name=atlantis version=0.27.2-r3 id=910a501f9587917b
time=2024-04-25T10:17:49.943-04:00 level=DEBUG msg="created Syft source from directory" description="{ID:b96c52ce347e32acf553fc32fadf9093f81e52990d4377509ad008623dceca33 Name:/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/wolfictl-sbom-1293666122 Version: Metadata:{Path:/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/wolfictl-sbom-1293666122 Base:}}"
time=2024-04-25T10:17:49.958-04:00 level=DEBUG msg="unpacked APK file to temp directory" apkFilePath=/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/x86_64-atlantis-0.27.2-r3-2452533565.apk
time=2024-04-25T10:17:49.958-04:00 level=DEBUG msg="apk temp directory item" path=.
time=2024-04-25T10:17:49.958-04:00 level=DEBUG msg="apk temp directory item" path=.PKGINFO
time=2024-04-25T10:17:49.958-04:00 level=DEBUG msg="apk temp directory item" path=.SIGN.RSA.wolfi-signing.rsa.pub
time=2024-04-25T10:17:49.958-04:00 level=DEBUG msg="apk temp directory item" path=usr
time=2024-04-25T10:17:49.958-04:00 level=DEBUG msg="apk temp directory item" path=usr/bin
time=2024-04-25T10:17:49.958-04:00 level=DEBUG msg="apk temp directory item" path=usr/bin/atlantis
time=2024-04-25T10:17:49.958-04:00 level=DEBUG msg="apk temp directory item" path=var
time=2024-04-25T10:17:49.958-04:00 level=DEBUG msg="apk temp directory item" path=var/lib
time=2024-04-25T10:17:49.958-04:00 level=DEBUG msg="apk temp directory item" path=var/lib/db
time=2024-04-25T10:17:49.961-04:00 level=DEBUG msg="apk temp directory item" path=var/lib/db/sbom
time=2024-04-25T10:17:49.961-04:00 level=DEBUG msg="apk temp directory item" path=var/lib/db/sbom/atlantis-0.27.2-r3.spdx.json
time=2024-04-25T10:17:49.962-04:00 level=DEBUG msg="synthesized APK package for SBOM" name=atlantis version=0.27.2-r3 id=05918cfba5895678
time=2024-04-25T10:17:49.962-04:00 level=DEBUG msg="created Syft source from directory" description="{ID:5c4f2ea53204320f2fd5adbde362288ffd2d0249e0534724c6700410d71e793c Name:/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/wolfictl-sbom-3770624166 Version: Metadata:{Path:/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/wolfictl-sbom-3770624166 Base:}}"
time=2024-04-25T10:17:50.700-04:00 level=DEBUG msg="finished Syft SBOM generation" packageCount=105
time=2024-04-25T10:17:50.700-04:00 level=DEBUG msg="cleaning up temp directory" path=/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/wolfictl-sbom-3770624166
time=2024-04-25T10:17:50.874-04:00 level=DEBUG msg="finished Syft SBOM generation" packageCount=105
time=2024-04-25T10:17:50.874-04:00 level=DEBUG msg="cleaning up temp directory" path=/var/folders/pz/h07rtjmd4mv4wpnr5st5q1480000gn/T/wolfictl-sbom-1293666122
time=2024-04-25T10:17:51.380-04:00 level=DEBUG msg="scanning APK SBOM for vulnerabilities" packageCount=105
time=2024-04-25T10:17:51.381-04:00 level=DEBUG msg="converted packages to grype packages" packageCount=105
time=2024-04-25T10:17:51.390-04:00 level=DEBUG msg="grype matching finished" matchCount=1
time=2024-04-25T10:17:51.390-04:00 level=DEBUG msg="scanning APK SBOM for vulnerabilities" packageCount=105
time=2024-04-25T10:17:51.391-04:00 level=DEBUG msg="converted packages to grype packages" packageCount=105
time=2024-04-25T10:17:51.401-04:00 level=DEBUG msg="grype matching finished" matchCount=1
time=2024-04-25T10:17:51.402-04:00 level=DEBUG msg="cleaned up after scan"
```

File: pkg/sbom/sbom.go [hunk 1]
```diff
@@ -78,6 +108,8 @@ func Generate(ctx context.Context, inputFilePath string, f io.Reader, distroID s
 	packageCollection := createdSBOM.Artifacts.Packages
 	packageCollection.Add(*apkPackage)
 
+	logger.Debug("finished Syft SBOM generation", "packageCount", packageCollection.PackageCount())
```
Comment (hectorj2f) [inline pkg/sbom/sbom.go:111]: I'd like to raise the level of this log entry for the moment to `Info` level.

Comment (hectorj2f) [reply pkg/sbom/sbom.go:111]: If we could add the package name too.

Comment (luhring) [reply pkg/sbom/sbom.go:111]: Good idea, will do!

File: pkg/scan/apk.go [hunk 2]
```diff
@@ -167,6 +159,8 @@ func (s *Scanner) APKSBOM(ssbom *sbomSyft.SBOM) (*Result, error) {
 	syftPkgs := ssbom.Artifacts.Packages.Sorted()
 	grypePkgs := grypePkg.FromPackages(syftPkgs, grypePkg.SynthesisConfig{GenerateMissingCPEs: false})
 
+	logger.Debug("converted packages to grype packages", "packageCount", len(grypePkgs))
```
Comment (hectorj2f) [inline pkg/scan/apk.go:162]: I'd also change this to Info too.

**Review by hectorj2f — DISMISSED** [review]
lgtm, minor comments.

File: pkg/scan/apk.go (hunk 2 above)
Comment (luhring) [reply pkg/scan/apk.go:162]: Sounds good!

**Review by hectorj2f — APPROVED** [review]

**Pull request merged**